apps/concierge_site/lib/controllers/auth_controller.ex

```diff
@@ -0,0 +1,105 @@
+defmodule ConciergeSite.AuthController do
+  use ConciergeSite.Web, :controller
+
+  require Logger
+
+  alias AlertProcessor.Model.User
+  alias AlertProcessor.Repo
+  alias ConciergeSite.SessionHelper
+  alias Plug.Conn
+
+  plug(Ueberauth)
+
+  @spec register(Conn.t(), map()) :: Conn.t()
+  def register(conn, _params) do
+    base_uri = System.get_env("KEYCLOAK_BASE_URI")
+    client_id = System.get_env("KEYCLOAK_CLIENT_ID")
+    redirect_uri = "KEYCLOAK_REDIRECT_URI" |> System.get_env() |> URI.encode()
+
+    registration_uri =
+      "#{base_uri}/auth/realms/MBTA/protocol/openid-connect/registrations?client_id=#{client_id}&response_type=code&scope=openid&redirect_uri=#{redirect_uri}"
+
+    redirect(conn, external: registration_uri)
+  end
+
+  @spec callback(Conn.t(), any) :: Conn.t()
+  def callback(
+        %{
+          assigns: %{
+            ueberauth_auth: %{
+              credentials: %{
+                other: %{
+                  user_info: %{"mbta_uuid" => id, "email" => email} = user_info
+                }
+              }
+            }
+          }
+        } = conn,
+        _params
+      ) do
+    phone_number =
+      user_info
+      |> Map.get("phone")
+      |> strip_us_country_code()
+
+    user =
+      id
+      |> get_or_create_user(email, phone_number)
+      |> use_email_and_phone_from_token(email, phone_number)
```

Comment:
**question:** just so I understand the transition plan correctly - it looks like you're using the email and phone number from the SSO response to create a new user if an old user isn't present, but otherwise you don't update those fields in the DB... though I'm guessing that since you're looking up by the ID in the SSO response this would _only_ give you new users? Or is the MBTA SSO ID specially created to bridge the old and new accounts?

Reply:
Yeah, good questions, and maybe they imply that I should add some comments or something to describe this better.

I'll start with the last part: the plan is to export all existing user accounts into Keycloak, which will include the existing IDs. So lookups should cover all existing users, whether they created their account pre- or post-Keycloak transition. The case that won't be found in the database is a user who just created their account in Keycloak and is arriving here for the first time. So those are the ones we need to write to the database.

Getting to the bigger question, we have a system where the user changes their account details (email, phone) in Keycloak, but since we need to save that data locally to be able to send notifications, Keycloak sends a message for each of those updates via SQS. T-Alerts will regularly check SQS for new messages, and apply the given change to the local database so that the local data mirrors the authoritative source in Keycloak. But there's a possible/likely scenario where a user goes to Keycloak to edit their account, and then bounces back to T-Alerts. So the data in the token reflects the latest updates, but it's likely that we haven't yet gotten the message and updated our own database, so our local data might show the old values. So we want to use the new values they just set for what we show on the page since those are accurate. But since we know that the SQS system will eventually update the values in or DB to match, we don't need to actually worry about writing to the database here.

Sorry, that's long. I'll try to put something more concise into a comment.